4-pknujsp

pknujsp opened the pull request:
<!-- PR은 최대한 다른 사람이 알아보기 쉽도록 자세히 써주세요. 특히 수도 코드 부분은 더더욱요...!!-->

## 🔗 문제 링크
<!-- 해결한 문제의 링크를 올려주세요. -->
[튜플](https://school.programmers.co.kr/learn/courses/30/lessons/64065)

## ✔️ 소요된 시간
<!-- 문제를 해결하는데 소요된 시간을 적어주세요. -->
30분

## ✨ 수도 코드
<!-- 내가 작성한 코드를 모르는 사람이 봐도 이해할 수 있도록 글로 쉽게 풀어서 설명해주세요. -->

- **E** : 입력 문자열의 각 원소
- **e** : E의 각 원소

1. 문자열로 원소 목록이 표현되어있으므로, 정규표현식으로 **E**를 각각 추출한다.
2. 튜플의 원소를 **key**, 배치 순서를 **value**로 하는 **Map**을 생성한다.
3. **E**를 순회하면서 아래 **4, 5**과정을 처리한다.
4.  **E**의 원소 **e**가 **Map**에 포함된 경우 -> **기존 값**과 **E의 크기** 중 **최솟값**으로 갱신
5. **e**가 **Map**에 없는 경우 -> **E의 크기**를 저장
6. 순회가 완료된 후, **Map**의 **entry**를 순회하면서 튜플을 완성한다.

### 3, 4번 과정

![](https://github.com/AlgoLeadMe/AlgoLeadMe-1/assets/48265129/9c865ad9-f2d6-4abf-94bd-5f70564d5a98)

**E**를 순회할 때 마다, **e**의 튜플 내 위치가 갱신된다.

1. #### `4, 2, 3`
2. 모두 첫 출현 -> 튜플 내 위치 **3**으로 설정
3. #### `3`
4. 재 출현 -> E의 크기가 1, 기존값은 3 -> 튜플에서 가장 앞에 있는 원소 -> 둘 중 최솟값인 **1**로 설정
5. #### `2, 3, 4, 1` 
6. `2, 4` 재 출현 -> 기존 값이 3이고 E의 크기는 4인데, 이는 `2, 3`중 하나가 크기가 **1~2**인 E에서 나옴을 의미(재 출현) -> 둘 중 최솟값인 **3**으로 갱신
7. `3` 재 출현 -> 기존 값이 1이므로 갱신X
8. `1` 첫 출현 -> **4**로 설정
9. #### `2, 3`
10. `2, 3` 모두 재 출현 -> `2`가 **E의 크기 2**로 갱신되고, `1`은 그대로 **1**로 유지

## 📚 새롭게 알게된 내용
<!-- 새롭게 알게된 내용이 있다면 작성 해주시고 출처를 남겨주세요. -->

Munbin-Lee commented:
```python
from collections import defaultdict

def solution(s):
    answer = []
    visited = defaultdict(bool)
    
    for i in sorted(s[2:-2].split('},{'), key=len):
        for j in map(int, i.split(',')):
            if visited[j]: continue
            visited[j] = True
            answer.append(j)
    return answer
```

예전에 C++로 풀다가 문자열 처리가 ```X```같아서 파이썬으로 풀었던 문제네요.

각 튜플을 길이순으로 오름차순 정렬한 후, 튜플 각 원소의 방문 여부 ```visited```를 측정하면서 ```answer```에 저장해요.

예를 들어, ```{{2},{1,2},{1,3,2},{4,1,3,2}}```로 정렬한 후, 2를 저장, 1을 저장, 3을 저장, 4를 저장해요.

pknujsp commented:
> ```python
> from collections import defaultdict
> 
> def solution(s):
>     answer = []
>     visited = defaultdict(bool)
>     
>     for i in sorted(s[2:-2].split('},{'), key=len):
>         for j in map(int, i.split(',')):
>             if visited[j]: continue
>             visited[j] = True
>             answer.append(j)
>     return answer
> ```
> 
> 예전에 C++로 풀다가 문자열 처리가 `X`같아서 파이썬으로 풀었던 문제네요.
> 
> 각 튜플을 길이순으로 오름차순 정렬한 후, 튜플 각 원소의 방문 여부 `visited`를 측정하면서 `answer`에 저장해요.
> 
> 예를 들어, `{{2},{1,2},{1,3,2},{4,1,3,2}}`로 정렬한 후, 2를 저장, 1을 저장, 3을 저장, 4를 저장해요.

확실히 파이썬은 사기적 이네요 ㄷㄷ

Munbin-Lee commented on pknujsp/SORT/4-튜플.kt:
```diff
@@ -0,0 +1,32 @@
+class Solution {
+    fun solution(s: String): IntArray {
+        // s를 목록화 하였을떄 원소의 정규표현식
+        val pattern = Regex("""\{{1,2}([0-9,]+)\}{1,2}""")
```
```{```를 1개 또는 2개 인식한 후, 숫자와 ```,```를 캡처하고, ```}```를 1개 또는 2개를 인식...

정규표현식 사용할 생각은 못했는데 깔끔하네요.

tgyuuAn replied:
정규표현식을 사용하다니 ㅎㄷㄷ;

Munbin-Lee commented on pknujsp/SORT/4-튜플.kt:
```diff
@@ -0,0 +1,32 @@
+                    lastV = elementMap[v]
+                    // 원소맵에 값이 있는 경우 -> 기존 값과 현재 원소목록의 크기 중 최솟값을 저장
+                    // 값이 없으면 -> 현재 원소목록의 크기를 저장
+                    elementMap[v] = if(lastV != null) minOf(elementsSize, lastV!!) else elementsSize
```
정렬할 필요 없이 그냥 최솟값 갱신ㄷㄷㄷㄷ

개천재다ㄷㄷㄷㄷ

Munbin-Lee reviewed (approved):
![LGTM](https://github.com/AlgoLeadMe/AlgoLeadMe-1/assets/100560031/ef282bca-bc0d-4940-b433-d2ec6952565a)

tgyuuAn reviewed (approved):
아니 풀이 진짜 창의적인데요 ??

개천재 ㅎㄷㄷ

tgyuuAn commented:
저도 문빈님이랑 로직은 같네요.

```python
def solution(s):
    s=s.strip("{")
    s=s.strip("}")
    s=s.split("},{")
    s.sort(key=lambda x : len(x))

    answer=[]
    for i in s:
        i=i.split(",")
        for j in i:
            if int(j) not in answer:
                answer.append(int(j))
    
    print(answer)
    
    return answer
```

마지막에 모집합을 만드는 과정에서 defaultdict를 사용하신 것과 저는 그냥 반복문에서 list에 들어있는지 아닌지 판별하는 것만 다르군요
<br><br><br><br><br><br><br><br><br><br><br><br>
그나저나 pknujsp 풀이 완전 참신하네요 ㅎㄷㄷ... 짱신기
